src/rust_helpers/weak_self.rs

```diff
@@ -0,0 +1,79 @@
+// copied from https://github.com/eun-ice/weak-self
+
+use std::cell::UnsafeCell;
+use std::fmt;
+use std::rc::{Rc, Weak};
+
+pub struct WeakSelf<T: ?Sized> {
+    cell: UnsafeCell<Option<Weak<T>>>,
+}
+
+impl<T: ?Sized> Clone for WeakSelf<T> {
+    fn clone(&self) -> Self {
+        WeakSelf {
+            cell: UnsafeCell::new(Some(self.get())),
+        }
+    }
+}
+
+impl<T: ?Sized> WeakSelf<T> {
+    /// Constructs a new empty WeakSelf<T>
+    pub fn new() -> WeakSelf<T> {
+        WeakSelf {
+            cell: UnsafeCell::new(None),
+        }
+    }
+
+    /// Initialize the WeakSelf<T> with an Rc.
+    ///
+    /// Note: content must point be the only existing Rc, otherwise this method will panic
+    pub fn init(&self, content: &Rc<T>, require_exclusive_access: bool) {
+        if require_exclusive_access {
```

Comment:
I modified `WeakSelf` to accept a flag for whether to panic if we don't have exclusive access to the `Rc` contents at the time of initialization because for the `true_type`/`regular_true_type` use case we've already initialized one weak reference to each before we can then initialize the other weak reference to them (since they both get referenced by themselves and each other)